src/main.cpp

```diff
@@ -4685,27 +4700,26 @@ bool AlreadyHaveBlock(const CInv &inv) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
     return true;
 }
 
-void static ProcessGetData(CNode *pfrom, const Consensus::Params &consensusParams)
+bool static ProcessGetData(CNode *pfrom, const Consensus::Params &consensusParams, std::deque<CInv> &vInv)
 {
-    std::deque<CInv>::iterator it = pfrom->vRecvGetData.begin();
+    bool gotWorkDone = false;
 
     std::vector<CInv> vNotFound;
 
-    LOCK(cs_main);
-
-    while (it != pfrom->vRecvGetData.end())
+    while (!vInv.empty())
     {
         // Don't bother if send buffer is too full to respond anyway
         if (pfrom->nSendSize >= SendBufferSize())
             break;
```

Comment:
Unless I'm mistaken nSendSize is neither locked nor atomic...is it possible multiple threads could be processing for the same node and  accessing this?

Reply:
This is a good thought.  In reviewing it, modifications are always locked by cs_vSend.  Reads are not locked.  But any read corruption would be innocuous (pop out of a loop 1 early or late) except that in 2 places we Misbehaving(50) the node.  Perhaps we should just misbehaving 10 rather than synchronizing every read?